### PR TITLE
[auto-materialize] Export `AutoMaterializeRule` class, add docs

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Allow `from dagster import AutoMaterializeRule`, and refresh the docs.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py [hunk 1]
```diff
@@ -69,30 +69,43 @@ class AutoMaterializePolicy(
...
-    For an asset / partition of an asset with a _lazy_ policy to be auto-materialized, at least one
-    of the following must be true:
+    In essence, the eager policy will cause an asset to be materialized whenever it is missing,
+    whenever any of its parent assets are updated, or whenever its freshness policy or that of one
+    of its downstream assets requires it to be updated in order to meet those policies, unless any
+    of its parent asset / partitions are missing, or are out of date with respect to their parents.
 
-    - it has a freshness policy that requires more up-to-date data
-    - any of its descendants have a freshness policy that require more up-to-date data
+    Policies can be customized, by adding or removing rules. For example, if you'd like to allow
```
Comment (erinkcochran87) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:89]:
```suggestion
    Policies can be customized by adding or removing rules. For example, if you'd like to allow
```

Nit - no comma needed here

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py [hunk 2]
```diff
@@ -69,30 +69,43 @@ class AutoMaterializePolicy(
 ):
     """An AutoMaterializePolicy specifies how Dagster should attempt to keep an asset up-to-date.
 
-    There are two main kinds of auto-materialize policies: eager and lazy. In essence, an asset with
-    an eager policy will try to immediately materialize after upstream changes, while an asset with
-    a lazy policy will only materialize when necessary in order to satisfy the relevant
-    FreshnessPolicies.
+    Each policy consists of a set of `AutoMaterializeRule`s, which are used to determine if there is
+    some reason that an asset / partition of an asset should be auto-materialized, and if so, if
+    there is some reason that an asset / partition of an asset should not be auto-materialized.
```
Comment (erinkcochran87) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:74]: I think we can restructure this to be easier to read - perhaps something like:

> Each policy consists of a set of `AutoMaterializeRule`s, which are used to determine whether an asset or a partition of an asset should or should not be auto-materialized.

Thoughts?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py [hunk 3]
```diff
@@ -69,30 +69,43 @@ class AutoMaterializePolicy(
...
 
-    For an asset / partition of an asset with an _eager_ policy to be auto-materialized, at least
-    one of the following must be true:
+    The most common policy is `AutoMaterializePolicy.eager()`, which consists of the following rules:
```
Comment (erinkcochran87) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:76]: I see you replaced the rule descriptions with the actual in-code rules, but I see value in keeping both. Thoughts on adding the original descriptions to their new in-code counterparts here? Ex:

> - `AutoMaterializeRule.materialize_on_missing()` - An asset or a partition of an asset is missing

Thoughts?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py [hunk 4]
```diff
@@ -69,30 +69,43 @@ class AutoMaterializePolicy(
...
-    - any of its parent assets / partitions have newer data
+    - `AutoMaterializeRule.materialize_on_missing()`
+    - `AutoMaterializeRule.materialize_on_parent_updated()`
+    - `AutoMaterializeRule.materialize_on_required_for_freshness()`
+    - `AutoMaterializeRule.skip_on_parent_outdated()`
+    - `AutoMaterializeRule.skip_on_parent_missing()`
 
-    For an asset / partition of an asset with a _lazy_ policy to be auto-materialized, at least one
-    of the following must be true:
+    In essence, the eager policy will cause an asset to be materialized whenever it is missing,
```
Comment (erinkcochran87) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:84]: If you add the descriptions alongside the rules above, I think you can nix this entire paragraph. IMO that approach makes this info easier to parse and results in (what feels like) less reading

**Comment by clairelin135** [comment]
@erinkcochran87 All of those suggestions make sense to me, I agree this seems more digestible. I've updated these descriptions, let me know what you think!

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py [hunk 5]
```diff
@@ -69,30 +69,42 @@ class AutoMaterializePolicy(
...
-    of the following must be true:
+    Each policy consists of a set of AutoMaterializeRules, which are used to determine whether an
+    asset or a partition of an asset should or should not be auto-materialized.
+
+    The most common policy is `AutoMaterializePolicy.eager()`, which consists of the following rules:
+
+    - `AutoMaterializeRule.materialize_on_missing()` - Materialize an asset or a partition if it has
+    never been materialized.
+    - `AutoMaterializeRule.materialize_on_parent_updated()` - Materialize an asset or a partition if
+    one of its parents has been updated more recently than it has.
```
Comment (erinkcochran87) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:80]: > one of its parents has been updated more recently than it has.

`has` should be `have` here. Weird ik, but it's referring to `parents` as the subject and not `asset` or `partition`

**Review by erinkcochran87 — APPROVED** [review]
Left a comment about one small thing I just noticed, but otherwise looks good to me!

**Pull request merged**